crates/qt-build-utils/src/builder/qmldir.rs

```diff
@@ -0,0 +1,145 @@
+// SPDX-FileCopyrightText: 2025 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
+// SPDX-FileContributor: Andrew Hayzen <andrew.hayzen@kdab.com>
+//
+// SPDX-License-Identifier: MIT OR Apache-2.0
+
+use crate::builder::qmluri::QmlUriBuilder;
+
+use std::io;
+
+/// QML module definition files builder
+///
+/// A qmldir file is a plain-text file that contains the commands
+pub struct QmlDirBuilder {
+    class_name: Option<String>,
+    plugin: Option<(bool, String)>,
+    type_info: Option<String>,
+    uri: QmlUriBuilder,
+}
+
+impl QmlDirBuilder {
+    /// Construct a [QmlDirBuilder] using the give [QmlUriBuilder] for the
+    /// module identifier
+    pub fn new(uri: QmlUriBuilder) -> Self {
+        Self {
+            class_name: None,
+            plugin: None,
+            type_info: None,
+            uri,
+        }
+    }
+
+    /// Writer the resultant qmldir text file contents
+    pub fn write(self, writer: &mut impl io::Write) -> io::Result<()> {
+        // Module is mandatory
+        let module = format!("module {}", self.uri.as_dots());
+
+        // Plugin, classname, and typeinfo are optional
+        let plugin = self
+            .plugin
+            .map(|(optional, name)| {
+                if optional {
+                    format!("optional plugin {name}")
+                } else {
+                    format!("plugin {name}")
+                }
+            })
+            .unwrap_or_default();
+        let class_name = self
+            .class_name
+            .map(|name| format!("classname {name}"))
+            .unwrap_or_default();
+        let type_info = self
+            .type_info
+            .map(|file| format!("typeinfo {file}"))
+            .unwrap_or_default();
+        let prefer = format!("prefer :/qt/qml/{}/", self.uri.as_dirs());
```

Comment:
As these are all just optionally appended to the writer, this should be refactored into a series of if statements:

```rust
if let Some(class_name) = self.class_name {
    write!(writer, "classname {class_name}");
}
```

That is IMO more readable and avoids a bunch of string allocations.